pyincore/models/hazard/tornado.py

```diff
@@ -0,0 +1,170 @@
+#
+# This program and the accompanying materials are made available under the
+# terms of the Mozilla Public License v2.0 which accompanies this distribution,
+# and is available at https://www.mozilla.org/en-US/MPL/2.0/
+from pyincore import HazardService, Dataset
+from pyincore.models.hazard.hazard import Hazard
+from pyincore.models.hazard.hazardDataset import TornadoDataset
+from pyincore.models.units import Units
+
+from shapely.geometry import Point
+import random
+
+
+class Tornado(Hazard):
+    """
+    "tornadoType": "dataset",
+    "id": "xxx",
+    "name": "pytest - Joplin Tornado",
+    "description": "Joplin tornado hazard",
+    "creator": "xxx",
+    "threshold": null,
+    "thresholdUnit": "mph",
+    "spaces": [
+        "xxx"
+    ],
+    "date": "2020-08-14T16:22:32+0000",
+    "datasetId": "xxx"
+    """
+
+    def __init__(self, metadata, ef_rating_field="ef_rating", ef_wind_speed=(65, 86, 111, 136, 166, 200),
+                 max_wind_speed=250.0, seed=1234):
+        super().__init__(metadata)
+        self.tornado_type = metadata["tornadoType"] if "tornadoType" in metadata else ""
+        # tornado has very different shape than other hazards
+        self.hazardDatasets = [
+            TornadoDataset({"threshold": metadata["threshold"] if "threshold" in metadata else None,
+                            "demandType": "wind",
+                            "demandUnits": metadata["thresholdUnit"] if "thresholdUnit" in metadata else "mph",
+                            "datasetId": metadata["datasetId"] if "datasetId" in metadata else ""})
+        ]
+        self.hazard_type = "tornado"
+        self.EF_RATING_FIELD = ef_rating_field
+        self.EF_WIND_SPEED = ef_wind_speed
+        self.MAX_WIND_SPEED = max_wind_speed
+        self.SEED = seed
+
+    @classmethod
+    def from_hazard_service(cls, id: str, hazard_service: HazardService):
+        """Get Hazard from hazard service, get metadata as well.
+
+        Args:
+            id (str): ID of the Hazard.
+            hazard_service (obj): Hazard service.
+
+        Returns:
+            obj: Hazard from Data service.
+
+        """
+        metadata = hazard_service.get_tornado_hazard_metadata(id)
+        instance = cls(metadata)
+        return instance
+
+    def read_hazard_values(self, payload: list, hazard_service=None, **kwargs):
+        """ Retrieve bulk earthquake hazard values either from the Hazard service or read it from local Dataset
+
+        Args:
+            payload (list):
+            hazard_service (obj): Hazard service.
+            kwargs (dict): Keyword arguments.
+        Returns:
+            obj: Hazard values.
+
+        """
+        if self.id and self.id != "" and hazard_service is not None:
+            return hazard_service.post_tornado_hazard_values(self.id, payload, **kwargs)
+        else:
+            return self.calculate_wind_speed_uniform_random_dist(payload)
+
+    def calculate_wind_speed_uniform_random_dist(self, payload):
```

Comment:
The default actually uses a Mersenne twister for random values. We'll need to see if python random has an implementation.

Reply:
> Almost all module functions depend on the basic function [random()](https://docs.python.org/3/library/random.html#random.random), which generates a random float uniformly in the half-open range 0.0 <= X < 1.0. Python uses the Mersenne Twister as the core generator. It produces 53-bit precision floats and has a period of 2**19937-1. The underlying implementation in C is both fast and threadsafe. The Mersenne Twister is one of the most extensively tested random number generators in existence.

does that mean it's already default to be Mersenne twister? https://docs.python.org/3/library/random.html#:~:text=Python%20uses%20the%20Mersenne%20Twister,random%20number%20generators%20in%20existence.

Reply:
Yes, the default in TornadoParameters on the service is Mersenne unless the user specifies to use the uniform distribution.

Reply:
The default in the service is to use a uniform real distribution using an apache commons math library. The idea is that the wind is random inside an EF zone, bounded by the wind speed for the EF box (e.g. EF1 is between 86 MPH and 111 MPH). That means at any point in the EF box a location could experience the maximum. You can see it here: https://github.com/IN-CORE/incore-services/blob/develop/server/hazard-service/src/main/java/edu/illinois/ncsa/incore/service/hazard/models/tornado/utils/TornadoCalc.java#L151

I think we need to check for the windSpeedMethod in the tornadoParameters and if it's 1 (default) we should use the uniform real distribution. The other option is what you have now where the speed increases linearly across the EF boundary so depending where you are in the box will impact what wind speed gets computed. I think they prefer the randomness for wind speed since that's more typical.

Reply:
Thanks for the information and the code. I think i understand a little bit more now.  Yes I can try the random uniform.